File: 진우/13weeks/12789. 도키도키 간식드리미/도키도키 간식드리미.py
N = int(input())  # 총 사람 수 입력

line = list(map(int, input().split()))  # 대기줄에 서 있는 사람들의 번호

stack = []         # 옆으로 빠지는 대기 공간(스택)
current = 1        # 지금 나와야 하는 번호 (1부터 시작)

for i in range(N):
    # 줄 맨 앞의 사람이 현재 나와야 하는 번호라면 바로 보냄
    if line[i] == current:
        current += 1
    else:
        # 아니라면 스택에 대기
        stack.append(line[i])

    # 스택의 맨 위에 있는 사람이 현재 나와야 하는 번호라면 계속 꺼냄
    while stack and current == stack[-1]:
        stack.pop()
        current += 1

# 스택이 비어있으면 모든 사람이 순서대로 나갔다는 뜻
if stack:
    print("Sad")
else:
    print("Nice")

File: 진우/13weeks/12789. 도키도키 간식드리미/도키도키 간식드리미.cc
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;

    stack<int> s;   // 대기열 역할 (옆으로 빼두는 공간)
    int current = 1; // 현재 순서대로 나와야 하는 번호 (1부터 시작)

    for (int i = 0; i < N; i++) {
        int num;
        cin >> num;

        // **지금 들어온 번호가 순서대로 나와야 한다면**
        if (num == current) {
            current++;  // 바로 통과
        } 
        else {
            // 순서가 아니라면 대기열(스택)에 보관
            s.push(num);
        }

        // **대기열에서 꺼낼 수 있을 때 반복해서 꺼내기**
        while (!s.empty() && s.top() == current) {
            s.pop();
            current++;
        }
    }

    // 대기열이 모두 비었으면 모든 번호가 순서대로 나왔다
    if (s.empty()) cout << "Nice" << "\n";
    else cout << "Sad" << "\n";

    return 0;
}
